In [46]:
import pandas as pd
from scipy.stats import chi2_contingency, mannwhitneyu

# Hypotéza 1
Na silnicích první třídy se při nehodách umíralo se stejnou pravděpodobností jako na silnicích třetí třídy.

Načteme data a zobrazíme tabulku

In [47]:
df = pd.read_pickle("accidents.pkl.gz")
df

p1  p36    p37         p2a  weekday(p2a)   p2b  p6  p7  p8  \
0       002100160001    4     -1  2016-01-01             5    55   1   1   0   
1       002100160002    4     -1  2016-01-01             5   130   1   3   0   
2       002100160003    5     -1  2016-01-01             5   100   1   2   0   
3       002100160004    6     -1  2016-01-01             5   120   9   0   0   
4       002100160005    6     -1  2016-01-01             5  2560   2   0   0   
...              ...  ...    ...         ...           ...   ...  ..  ..  ..   
572929  190906210457    6     -1  2021-08-30             1   506   2   0   0   
572930  190906210458    6     -1  2021-08-30             1   752   2   0   0   
572931  190906210459    3   2099  2021-08-30             1  1505   1   3   0   
572932  190906210460    3   2099  2021-08-31             2  1245   0   0   0   
572933  190906210461    3  21029  2021-08-31             2  1834   3   0   4   

        p9  ...      l        n  o                       p       q       r  \
0        2  ...          711403     Souhlasnýsesměremúseku  Pomalý  554782   
1        2  ...                                                         -1   
2        2  ...                                                         -1   
3        2  ...                                                         -1   
4        2  ...                                                         -1   
...     ..  ...    ...      ... ..                     ...     ...     ...   
572929   2  ...         2611387                             Pomalý  560359   
572930   2  ...         2541503         Opačnýkesměruúseku  Pomalý  560383   
572931   2  ...                                                         -1   
572932   1  ...   2099   199764         Opačnýkesměruúseku  Pomalý  560286   
572933   1  ...  21029  3650728     Souhlasnýsesměremúseku  Pomalý  560324   

             s                   t  p5a  region  
0       451622  GN_V0.1UIR-ADR_410    1     PHA  
1           -1                        1     PHA  
2           -1                        1     PHA  
3           -1                        1     PHA  
4           -1                        1     PHA  
...        ...                 ...  ...     ...  
572929      -1  GN_V0.1UIR-ADR_410    1     KVK  
572930  550426  GN_V0.1UIR-ADR_410    1     KVK  
572931      -1                        1     KVK  
572932  553654  GN_V0.1UIR-ADR_410    1     KVK  
572933      -1  GN_V0.1UIR-ADR_410    2     KVK  

[572934 rows x 65 columns]

Rozhodneme, u kterých nehod došlo k fatálním následkům a omezíme rozsah na silnice 1. a 3. třídy

In [48]:
df2 = df[(df["p36"].isin([1, 3]))].copy()
df2["death"] = df2["p13a"] > 0
df2

p1  p36    p37         p2a  weekday(p2a)   p2b  p6  p7  p8  \
7594    002100167619    1     -1  2016-05-05             4  1135   1   2   0   
10530   003100160575    1    102  2016-06-19             0  1905   1   2   0   
10532   003100160577    1      0  2016-06-19             0  1915   3   0   4   
13362   003100163418    1      0  2016-08-08             1  1355   1   2   0   
21478   004100161576    1     -1  2016-12-09             5  1640   2   0   0   
...              ...  ...    ...         ...           ...   ...  ..  ..  ..   
572914  190906210442    3  21036  2021-08-23             1  1238   3   0   9   
572922  190906210450    3  21810  2021-08-26             4  1704   1   1   0   
572931  190906210459    3   2099  2021-08-30             1  1505   1   3   0   
572932  190906210460    3   2099  2021-08-31             2  1245   0   0   0   
572933  190906210461    3  21029  2021-08-31             2  1834   3   0   4   

        p9  ...        n  o                       p       q       r       s  \
7594     2  ...  1397508     Souhlasnýsesměremúseku  Pomalý  554782  709450   
10530    2  ...   312370     Souhlasnýsesměremúseku  Pomalý  554782  481548   
10532    2  ...   293620     Souhlasnýsesměremúseku  Pomalý  554782  709450   
13362    2  ...  2374326     Souhlasnýsesměremúseku  Pomalý  554782  506001   
21478    2  ...                                                  -1      -1   
...     ..  ...      ... ..                     ...     ...     ...     ...   
572914   1  ...  2245808         Opačnýkesměruúseku  Pomalý  560341      -1   
572922   2  ...  3343415     Souhlasnýsesměremúseku  Pomalý  560677      -1   
572931   2  ...                                                  -1      -1   
572932   1  ...   199764         Opačnýkesměruúseku  Pomalý  560286  553654   
572933   1  ...  3650728     Souhlasnýsesměremúseku  Pomalý  560324      -1   

                         t  p5a  region  death  
7594    GN_V0.1UIR-ADR_410    1     PHA  False  
10530   GN_V0.1UIR-ADR_410    1     PHA  False  
10532   GN_V0.1UIR-ADR_410    1     PHA  False  
13362   GN_V0.1UIR-ADR_410    1     PHA  False  
21478   GN_V0.1UIR-ADR_410    1     PHA  False  
...                    ...  ...     ...    ...  
572914  GN_V0.1UIR-ADR_410    2     KVK  False  
572922  GN_V0.1UIR-ADR_410    1     KVK  False  
572931                        1     KVK  False  
572932  GN_V0.1UIR-ADR_410    1     KVK  False  
572933  GN_V0.1UIR-ADR_410    2     KVK  False  

[153329 rows x 66 columns]

Vytvoříme kontingenční tabulku, která zobrazuje počet smrtelných nehod a nehod bez umrtí na silnicích 1. a 3. třídy.

In [49]:
ct = pd.crosstab(df2['death'], df2['p36'])
ct

p36        1      3
death              
False  78618  73352
True     911    448

Nyní s pomocí **$\chi^2$ chí-kvadrát testu** ověříme, jestli existuje mezi proměnnými statisticky významná korelace.

In [50]:
result = chi2_contingency(ct)
result[1]

3.5395243450138555e-29

Hodnota $3.54\cdot10^{-29}$ je mnohem menší než $0.05$. Proto zamítneme nulovou hypotézu.

Očekávané frekvence:

In [51]:
result[3]

array([[78824.11109444, 73145.88890556],
       [  704.88890556,   654.11109444]])

In [52]:
ct - result[3]

p36             1           3
death                        
False -206.111094  206.111094
True   206.111094 -206.111094

Z výstupu $\chi^2$ chí-kvadrát testu je také vidět, že nehody na silnicích 1. třídy vedly častěji ke smrtelným zraněním.

# Hypotéza 2

Při nehodách vozidel značky Škoda je škoda na vozidle nižší než při nehodách vozidel Audi.

Nejprve vyfiltrujeme vozy Audi (2) a Škoda (39) a vyčíslení opravy.

In [53]:
# 2 == Audi, 39 == Skoda
df3 = df[df["p45a"].isin([2, 39])].copy()
df3 = df3[["p45a", "p53"]]
df3

p45a   p53
0          2  4000
6         39   200
9         39   300
11        39    50
16        39  1200
...      ...   ...
572911    39    10
572916    39   130
572921    39   100
572923    39     0
572931     2    50

[129824 rows x 2 columns]

Dále vypočítáme porovnání dvou konfigurací (Audi a Škoda), porovnáme mediány a spočítáme výsledek MWU-testu.

In [55]:
df4 = df3.set_index("p45a")
a = df4.loc[2, "p53"]
b = df4.loc[39, "p53"]

print(a.median(), b.median())
res = mannwhitneyu(a, b)
res

300.0 200.0


MannwhitneyuResult(statistic=789001736.0, pvalue=4.921643600739696e-188)

Z výsledku je vidět, že `pvalue` je mnohem menší než stanovená hodnota `0.05`. To znamená, že jsme testem vyvrátili podobnost v ceně opravy po nehodě u aut značky Škoda a Audi. Tomuto odpovídají i porovnané mediány, ze kterých je vidět, že opravy aut Audi jsou dražší.